### ISSUE TITLE
Flopping a List returns an Array

### ISSUE THREAD

**Issue by mtmccrea** [supercollider/supercollider#3591]
In trying to find a workaround for #3454 (`Array:flop` fails with large array sizes), I'm trying to force `SequenceableCollection:flop`, which works on large collections, by using `List` instead of `Array`. It's easy to miss that flopping a List returns an Array... So normal use cases it appears to work when flopping once, but if you flop again, as is done in [Matrix:with](https://github.com/supercollider-quarks/MathLib/blob/9927c0bd7bc5ce0189309321edfae5735bc4a14d/classes/various/Matrix.sc#L15), you'll again be calling `Array:flop`.

Assuming this is unintended behavior, the problem looks like this:

`SequenceableCollection:flop` will return a collection, which is created in the method using `this.species.fill()`:
https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/SCClassLibrary/Common/Collections/SequenceableCollection.sc#L486

`this.species` for `List` falls back on `Collection:species` which returns `Array`:
https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/SCClassLibrary/Common/Collections/Collection.sc#L89

I'm unfamiliar with "species" should return, I can't say if this is correct behavior.  `Set`, for example, returns `this.class` from `.species`:
https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/SCClassLibrary/Common/Collections/Set.sc#L5

Changing `Collection:species` to return `this.class`, seems to fix my immediate problem, but I'm thinking this isn't the best approach.

So a couple takeaway questions: 
*Should `Collection:species` return `Array`?
*What's the difference between `.species` and `.class`, given their different implementations in different types of collections?

Thanks!

**Comment by mtmccrea** [comment]
ping @joslloand

**Comment by telephon** [comment]
The reason for `species` is that it can be much more efficient to have arrays here. Also it is slightly unclear whether you want lists in lists or or arrays in a list.

**Comment by mtmccrea** [comment]
Thanks for the clarification about efficiency. Do you mean that the purpose of “species” is to force the object to become an array, making the assumption of certain use cases for the method? 

Ultimately I would expect that doing an operation on some kind of collection would return a collection of the original type. 

I see your point that it gets more complicated with a list of arrays. In that case I suppose flopping a list of arrays would return a list of arrays.

**Comment by telephon** [comment]
Yes, we've had some discussions in the past about this, and my initial intuition was just like yours.

Here is the list of what is the species:

```
Event -> Event
Environment -> Environment
FuzzySet -> FuzzySet
FuzzyDictionary -> FuzzyDictionary
ProxyNodeMap -> ProxyNodeMap
NodeMap -> NodeMap
IdentityDictionary -> IdentityDictionary
Dictionary -> Dictionary
OrderedIdentitySet -> OrderedIdentitySet
IdentitySet -> IdentitySet
Set -> Set
Range -> Range
Archive -> Archive
Library -> Library
LibraryBase -> LibraryBase
MultiLevelIdentityDictionary -> MultiLevelIdentityDictionary
Interval -> Interval
ObjectTable -> ObjectTable
TwoWayIdentityDictionary -> TwoWayIdentityDictionary
Array2D -> Array2D
Pair -> Pair
IdentityBag -> IdentityBag
Bag -> Bag
RingBuffer -> RingBuffer
LinkedList -> LinkedList
SparseArray -> SparseArray
Order -> Order
SortedList -> SortedList
ClearList -> ClearList
VagueList -> VagueList
List -> List
Polynomial -> Polynomial
DoubleArray -> DoubleArray
Wavetable -> Wavetable
Signal -> Signal
FloatArray -> FloatArray
String -> String
SymbolArray -> SymbolArray
Int32Array -> Int32Array
Int16Array -> Int16Array
Int8Array -> Int8Array
RawArray -> RawArray
SubOrd -> SubOrd
Ord -> Ord
NFunc -> NFunc
SystemNFunc -> SystemNFunc
Matrix -> Matrix
Array -> Array
ArrayedCollection -> ArrayedCollection
SequenceableCollection -> SequenceableCollection
Collection -> Collection
```

**Comment by mtmccrea** [comment]
It looks like from your output `this.species` is identical to `this.class` (I'm not sure how you generated this output)
I see a discrepancy in my results:
```
List.with(1,2,3,4).species
-> Array
```
What am I missing?

A quick look at all of the implementations of `-species` shows it returns `this.class` in every instance by `Collection`, which makes me think this must have been intentional. Presumably for efficiency reasons, as you mentioned?

So, I'm still not sure if this is "correct" behavior for `List`. If not, maybe adding 
```species { ^this.class}``` 
to `List` would be appropriate? Perhaps I should search the list for the previous discussion you mentioned...

**Comment by telephon** [comment]
hm, seem all to be the same.

**Comment by telephon** [comment]
ah I see. wait.

**Comment by telephon** [comment]
I'll do it tomorrow.

**Comment by telephon** [comment]
Originally, I had the same intuition like you. But the point is that sometimes other classes as intermediaries are better suited or needed.

Here is a listing of the species:

```
Event -> Event
Environment -> Environment
ProxyNodeMap -> ProxyNodeMap
NodeMap -> NodeMap
IdentityDictionary -> IdentityDictionary
Dictionary -> Dictionary
OrderedIdentitySet -> OrderedIdentitySet
IdentitySet -> IdentitySet
Set -> Set
Range -> Array
Archive -> Array
Library -> Array
LibraryBase -> Array
MultiLevelIdentityDictionary -> Array
Interval -> Array
ObjectTable -> Array
TwoWayIdentityDictionary -> Array
Array2D -> Array
Pair -> Array
IdentityBag -> Array
Bag -> Array
RingBuffer -> Array
LinkedList -> LinkedList
SparseArray -> SparseArray
Order -> Order
SortedList -> Array
List -> Array
DoubleArray -> Array
Signal -> Array
FloatArray -> Array
String -> String
SymbolArray -> Array
Int32Array -> Array
Int16Array -> Array
Int8Array -> Array
RawArray -> Array
Array -> Array
ArrayedCollection -> Array
SequenceableCollection -> Array
Collection -> Array
```

**Comment by telephon** [comment]
We need to review the implicit decisions in this list.

**Comment by telephon** [comment]
I wouldn't consider this a bug. It is a design decision which we'll have to carefully review at some time. Should we add it to a backlog?

**Comment by mtmccrea** [comment]
I agree it isn't necessarily a bug, but could use a review/reconsideration. In cases where the distinction between `List` and `Array` is important, it seems appropriate that calling a method on one wouldn't force you to then (perhaps unknowingly) be using the other.

Adding it to a backlog sounds fine to me.
